### PR TITLE
Add handling of MEMKIND_MEM_THRESHOLDS env

### PR DESCRIPTION
**Pull request by bratpiorka**
Add handling of MEMKIND_MEM_THRESHOLDS env

* this env will store thresholds configuration for DYNAMIC_THRESHOLD policy
* each threshold configuration query could contain VAL, MIN and MAX parameters
* added tests

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/641)

### CODE REVIEW

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @bratpiorka)

___
*[tiering/ctl.c, line 394 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PhvSbCD2z-MDQxCJp:-M_PhvSbCD2z-MDQxCJq:b-tls1m5) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L394)):*
> ```C
>     qbuf = strtok_r(qbuf, CTL_VALUE_SEPARATOR, &sptr);
>     if (!strcmp(qbuf, "POLICY")) {
>         ret = ctl_parse_policy(sptr, &policy);
> ```

NOTE that ctl_parse_policy() already prints err message in case of error 
___
*[tiering/ctl.c, line 403 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_Pi6GlDjHTAiKnfSWA:-M_Pi6GlDjHTAiKnfSWB:b-sgpkf3) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L403)):*
> ```C
>     }
> 
>     struct memtier_builder *builder = memtier_builder_new(policy);
> ```

NOTE that memtier_builder_new() already prints err message in case of error 
___
*[tiering/ctl.c, line 409 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiHc-DzkfNxa3-bWr:-M_PiHc-DzkfNxa3-bWs:b-ejly0n) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L409)):*
> ```C
> 
>     for (i = 0; i < tier_count; ++i) {
>         ret = memtier_builder_add_tier(builder, temp_cfg[i].kind,
> ```

NOTE that memtier_builder_add_tier() already prints err message in case of error 
___
*[tiering/ctl.c, line 456 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiQgM82I5RsxWur8a:-M_PiQgM82I5RsxWur8b:b-8yq6t2) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L456)):*
> ```C
>                 snprintf(buf, 128,
>                          "policy.dynamic_threshold.thresholds[%d].val", i);
>                 ret = memtier_ctl_set(builder, buf, (void *)&val);
> ```

I used here memtier_ctl_set because there is no access to builder members from this file. What is more, this function takes care of adding new threshold to builder and does all stuff related to default parameters if needed

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 5 files reviewed, 5 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[tiering/memtier.c, line 113 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiqfW3QDQAXC6TgHH:-M_Q07yr1CbUM9Dx8f1Z:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/memtier.c#L113)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> MEMKIND_MEM_THRESHOLDS
> ```
extend doc - therefore would be easier to review expected behavior
</blockquote></details>

Done.

**Review by michalbiesek — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 5 files reviewed, 11 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[tiering/ctl.c, line 272 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q2v35-5hnFPqY6_SP:-M_Q2v35-5hnFPqY6_SQ:b-pomu9z) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/ctl.c#L272)):*
> ```
> fsdax_size_str
> ```
wrong merge?
___
*[tiering/ctl.c, line 484 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q4LxI8LBjwbj6ciue:-M_Q4LxI8LBjwbj6ciuf:b-5bqato) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/ctl.c#L484)):*
> ```
> 128
> ```
64 should be sufficient could you make this as some define?
___
*[tiering/ctl.c, line 488 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q50SH6hRti2bKJfII:-M_Q50SI0jIVMupKVQuF:b-k0ti4h) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/ctl.c#L488)):*
> ```
> (void *)
> ```
not needed I believe
___
*[tiering/ctl.c, line 497 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q6AqI34HbZ5iTuUY8:-M_Q6AqI34HbZ5iTuUY9:b-1okkqf) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/ctl.c#L497)):*
> ```
> (void *)
> ```

not needed I believe
___
*[tiering/ctl.c, line 506 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q6CN89jrBZPJviWrx:-M_Q6CN89jrBZPJviWry:b-k0ti4h) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/ctl.c#L506)):*
> ```
> (void *)
> ```
not needed I believe
___
*[tiering/memtier.c, line 113 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q89gh30Hi3HSZrGAE:-M_Q89gh30Hi3HSZrGAF:b-clfijh) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/memtier.c#L113)):*
> ```
> char *thresholds_var = utils_get_env("MEMKIND_MEM_THRESHOLDS");
> ```
can this be called inside ctl_create_tier_memory_from_env? without extension of params
___
*[tiering/memtier.c, line 118 at r2](https://reviewable.io/reviews/memkind/memkind/641#-M_Q8N4TBoQI8Ke7_7dX:-M_Q8N4TBoQI8Ke7_7dY:b-wcggpu) ([raw file](https://github.com/memkind/memkind/blob/2468478e7c56526f485da5fe2aa496ea9539343d/tiering/memtier.c#L118)):*
> ```C
>             ctl_create_tier_memory_from_env(tiers_var, thresholds_var);
>         if (current_memory) {
>             return;
> ```

nit: commit message please use "-" instead of "*" and remove  
"added tests updated man" - which is obvious here.
Consider that you read this message as the first step to understand the commit - so it should point some information to better understand the changes from commit

**Review by PatKamin — COMMENTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[tiering/ctl.c, line 429 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_Q9QI785jZolBeoX6C:-M_Q9QI785jZolBeoX6D:b-yen7ua) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L429)):*
>= tier_count
___
*[tiering/ctl.c, line 444 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_QF3vQ6d3Y91fP8vb_:-M_QF3vQ6d3Y91fP8vba:b1vqt8j) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L444)):*
> ```C
>             size_t val = -1;
>             size_t min = -1;
>             size_t max = -1;
> ```

These values are wrapped around implicitly to the SIZE_MAX value. Therefore, following if statements will always be true.

**Review by michalbiesek — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 5 files reviewed, 6 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[man/memtier.7, line 26 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb3P_BaTauDsypdkw:-M_Qb3P_BaTauDsypdkx:bsimxes) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L26)):*
> ```
> MEMKIND_MEM_THRESHOLDS
> ```
This should be defined after MEMKIND_MEM_TIERS right?
___
*[man/memtier.7, line 62 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_QX6ESEasG_ZURX1PG:-M_QX6ESEasG_ZURX1PH:byfm524) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L62)):*
> ```
> provided memory kinds. This parameter has to be the last parameter in configuration
> string. Currently only
> ```
"Last parameter" in whole environment string or last parameter or in TIER PARAMETER NAME?'
___
*[man/memtier.7, line 83 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb9eB5_ISqGQLptCT:-M_Qb9eB5_ISqGQLptCU:b-ma2ncp) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L83)):*
> ```
> (optional)
> ```
there should be information about default values for each of the optional values - be aware that we assume that this will be an official interface - only memtier.7
___
*[tiering/ctl.c, line 403 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_Pi6GlDjHTAiKnfSWA:-M_QUyE58bYWbobtuMQV:biucdj7) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L403)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

NOTE that memtier_builder_new() already prints err message in case of error 
</blockquote></details>

Memtier_builder_new is currently part of the memkind module and to see:

log_err("Unrecognized memory policy %u", policy);

You have to define MEMKIND_DEBUG=1. The tiering/ctl.c have an independent logging mechanism so please revert the changes.
___
*[tiering/ctl.c, line 409 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiHc-DzkfNxa3-bWr:-M_QVLYL8ozg2tQA-R7M:bhlou83) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L409)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

NOTE that memtier_builder_add_tier() already prints err message in case of error 
</blockquote></details>

Similar situation as in the case of memtier_builder_new
___
*[tiering/ctl.c, line 456 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiQgM82I5RsxWur8a:-M_QXWH_CYtFczg3JIOx:b2ynws9) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L456)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

I used here memtier_ctl_set because there is no access to builder members from this file. What is more, this function takes care of adding new threshold to builder and does all stuff related to default parameters if needed 
</blockquote></details>

This is what kind of issue I try to address in #633
The CTL should have only one responsibility - setting one specific parameter - no more no less - no modifying the "max/min" when setting "val" no realloc inside. I am aware of my first approach to the independent order of the operations (like ctl and adding tier) and I admit it was wrong.
The order should be like in your implementation:
- create builder with specific policy - which is already done
- add tiers - which is already done
- define the additional parameter of the builder - in this case, threshold configuration - this is an interesting part :)

My opinion is like this the logic here should always call memtier_ctl_set:
 - if a situation when the parameter is defined in the env variable that's clear.
 - if some parameter is not defined the default values should be set here.

In other words, the memkind_memtier layer should not intervene by defaulting the parameters.

For now please, Since you support min/max/val here please:
add TODO for updating other parameters supported by CTL and provide doc for this. 
add TODO with reference to #633 that after a change from this PR the defaulting part should be changed as well and moved from memkind_memtier layer to this layer.

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[man/memtier.7, line 26 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb3P_BaTauDsypdkw:-M_UW2zq7bezYujVIfsb:b8p5syk) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L26)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> MEMKIND_MEM_THRESHOLDS
> ```
This should be defined after MEMKIND_MEM_TIERS right?
</blockquote></details>

I don't understand your comment. You mean order of env vars? They could be passed in any order.
___
*[man/memtier.7, line 62 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_QX6ESEasG_ZURX1PG:-M_USgUfELfwgR1SWrvv:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L62)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> provided memory kinds. This parameter has to be the last parameter in configuration
> string. Currently only
> ```
"Last parameter" in whole environment string or last parameter or in TIER PARAMETER NAME?'
</blockquote></details>

Done.
___
*[man/memtier.7, line 83 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb9eB5_ISqGQLptCT:-M_UevS76VF8JJJwculR:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L83)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> (optional)
> ```
there should be information about default values for each of the optional values - be aware that we assume that this will be an official interface - only memtier.7
</blockquote></details>

Done.
___
*[tiering/ctl.c, line 456 at r1](https://reviewable.io/reviews/memkind/memkind/641#-M_PiQgM82I5RsxWur8a:-M_UpH1UCRkH5EfDnhKY:b-7tfmjp) ([raw file](https://github.com/memkind/memkind/blob/9c49c7c6c9c5758e2d72e1542ebaa9827631325a/tiering/ctl.c#L456)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

This is what kind of issue I try to address in #633
The CTL should have only one responsibility - setting one specific parameter - no more no less - no modifying the "max/min" when setting "val" no realloc inside. I am aware of my first approach to the independent order of the operations (like ctl and adding tier) and I admit it was wrong.
The order should be like in your implementation:
- create builder with specific policy - which is already done
- add tiers - which is already done
- define the additional parameter of the builder - in this case, threshold configuration - this is an interesting part :)

My opinion is like this the logic here should always call memtier_ctl_set:
 - if a situation when the parameter is defined in the env variable that's clear.
 - if some parameter is not defined the default values should be set here.

In other words, the memkind_memtier layer should not intervene by defaulting the parameters.

For now please, Since you support min/max/val here please:
add TODO for updating other parameters supported by CTL and provide doc for this. 
add TODO with reference to #633 that after a change from this PR the defaulting part should be changed as well and moved from memkind_memtier layer to this layer.
</blockquote></details>

added

**Review by michalbiesek — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r4, 2 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 3 of 6 files reviewed, 2 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[man/memtier.7, line 26 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb3P_BaTauDsypdkw:-M_UqF7r73190lFkQppE:bxetb4c) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L26)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

I don't understand your comment. You mean order of env vars? They could be passed in any order.
</blockquote></details>

Any chance for the test which does configuration like MEMKIND_MEM_THRESHOLD then MEMKIND_MEM_TIERS?
___
*[tiering/ctl.c, line 369 at r7](https://reviewable.io/reviews/memkind/memkind/641#-M_UqlpU0LavQ8ZapGuj:-M_UqlpU0LavQ8ZapGuk:b-lpifc) ([raw file](https://github.com/memkind/memkind/blob/810fb7aea2200656dc90a4b32fde70b5afb4cbfa/tiering/ctl.c#L369)):*
> ```
> &st_min
> ```
I could be missing something but why you need to introduce new variable for setting this? (in all three cases)

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 2 of 6 files reviewed, 2 unresolved discussions (waiting on @michalbiesek and @PatKamin)

___
*[man/memtier.7, line 26 at r5](https://reviewable.io/reviews/memkind/memkind/641#-M_Qb3P_BaTauDsypdkw:-M_Ux17i6kVxkVWVGwwc:b-7tfmjp) ([raw file](https://github.com/memkind/memkind/blob/f9ebb894ee6279f2bebc1d2b4b48a794bd72fcab/man/memtier.7#L26)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

Any chance for the test which does configuration like MEMKIND_MEM_THRESHOLD then MEMKIND_MEM_TIERS?
</blockquote></details>

added
___
*[tiering/ctl.c, line 369 at r7](https://reviewable.io/reviews/memkind/memkind/641#-M_UqlpU0LavQ8ZapGuj:-M_UuL_-BFYeb8tLmrCK:bkmruek) ([raw file](https://github.com/memkind/memkind/blob/810fb7aea2200656dc90a4b32fde70b5afb4cbfa/tiering/ctl.c#L369)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> &st_min
> ```
I could be missing something but why you need to introduce new variable for setting this? (in all three cases)
</blockquote></details>

sorry, this was left from previous version of my code - removed

**Review by michalbiesek — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: 2 of 6 files reviewed, all discussions resolved (waiting on @michalbiesek and @PatKamin)

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 2 of 3 files at r3, 2 of 3 files at r7, 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bratpiorka)

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 4 of 4 files at r9.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bratpiorka)

**Review by PatKamin — COMMENTED** [review]
> Reviewed 2 of 3 files at r3, 2 of 3 files at r7, 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: all files reviewed, 1 unresolved discussion (waiting on @bratpiorka)

___
*[test/tiering_tests.py, line 498 at r8](https://reviewable.io/reviews/memkind/memkind/641#-M_Vi2BF81DNkCYfA0iV:-M_Vi2BGDzsAua9GYFT3:b-obgvj7) ([raw file](https://github.com/memkind/memkind/blob/3f36321d596eb5b8cafff510ad1a41b281d75eac/test/tiering_tests.py#L498)):*
> ```Python
>             negative_test=True)
> 
>     def test_negative_too_many_thresholds(self):
> ```

Please, extend this test case with parameter where there will be added each of these three values separately.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641)* status: all files reviewed, 1 unresolved discussion (waiting on @PatKamin)

___
*[test/tiering_tests.py, line 498 at r8](https://reviewable.io/reviews/memkind/memkind/641#-M_Vi2BF81DNkCYfA0iV:-M_ZSPgT1r4NzCh2vyk2:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/3f36321d596eb5b8cafff510ad1a41b281d75eac/test/tiering_tests.py#L498)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Please, extend this test case with parameter where there will be added each of these three values separately.
</blockquote></details>

Done.

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> Reviewed 3 of 4 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/641#-:-M_Ze59O-RT0OjdyvPql:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @bratpiorka)
